File: agent/tests/contract_test.rs
use edgemon_agent::protocol::*;
use std::fs;
use std::path::PathBuf;

fn fixture_path(name: &str) -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .join("protocol")
        .join("fixtures")
        .join(name)
}

#[test]
fn test_parse_hello_fixture() {
    let path = fixture_path("hello.json");
    let content = fs::read_to_string(&path).expect("Failed to read hello.json fixture");
    let envelope: Envelope<HelloData> =
        serde_json::from_str(&content).expect("Failed to parse hello.json");

    assert_eq!(envelope.v, 1);
    assert_eq!(envelope.msg_type, "hello");
    assert_eq!(envelope.data.agent.version, "0.1.0");
    assert_eq!(envelope.data.environment.env_type, "container");
    assert_eq!(envelope.data.resources.cpu_capacity_cores, Some(0.5));
    assert_eq!(envelope.data.resources.memory_limit_bytes, Some(536870912));
    assert_eq!(envelope.data.resources.rootfs_limit_bytes, None);
    assert_eq!(envelope.data.resources.rootfs_scope, "unknown");
}

#[test]
fn test_parse_welcome_fixture() {
    let path = fixture_path("welcome.json");
    let content = fs::read_to_string(&path).expect("Failed to read welcome.json fixture");
    let envelope: Envelope<WelcomeData> =
        serde_json::from_str(&content).expect("Failed to parse welcome.json");

    assert_eq!(envelope.v, 1);
    assert_eq!(envelope.msg_type, "welcome");
    assert_eq!(envelope.data.config_rev, 1);
    assert_eq!(envelope.data.config.sample_interval_sec, 2);
    assert_eq!(envelope.data.config.stream_interval_sec, 2);
    assert_eq!(envelope.data.config.probes.len(), 1);
    assert_eq!(envelope.data.config.probes[0].id, "cf-dns");
}

#[test]
fn test_parse_report_fixture() {
    let path = fixture_path("report.json");
    let content = fs::read_to_string(&path).expect("Failed to read report.json fixture");
    let envelope: Envelope<ReportData> =
        serde_json::from_str(&content).expect("Failed to parse report.json");

    assert_eq!(envelope.v, 1);
    assert_eq!(envelope.msg_type, "report");
    assert_eq!(envelope.data.cpu.usage_pct, Some(14.5));
    assert_eq!(envelope.data.memory.used_bytes, Some(184549376));
    assert_eq!(envelope.data.rootfs.used_bytes, None);
    assert_eq!(envelope.data.network.rx_total_bytes, 918273645);
    assert_eq!(envelope.data.probes.len(), 1);
    assert_eq!(envelope.data.probes[0].status, "ok");
}

#[test]
fn test_parse_config_fixture() {
    let path = fixture_path("config.json");
    let content = fs::read_to_string(&path).expect("Failed to read config.json fixture");
    let envelope: Envelope<ConfigData> =
        serde_json::from_str(&content).expect("Failed to parse config.json");

    assert_eq!(envelope.v, 1);
    assert_eq!(envelope.msg_type, "config");
    assert_eq!(envelope.data.config_rev, 2);
    assert_eq!(envelope.data.config.stream_interval_sec, 2);
    assert_eq!(envelope.data.config.probes.len(), 1);
}

#[test]
fn test_parse_config_ack_fixture() {
    let path = fixture_path("config_ack.json");
    let content = fs::read_to_string(&path).expect("Failed to read config_ack.json fixture");
    let envelope: Envelope<ConfigAckData> =
        serde_json::from_str(&content).expect("Failed to parse config_ack.json");

    assert_eq!(envelope.v, 1);
    assert_eq!(envelope.msg_type, "config_ack");
    assert_eq!(envelope.data.config_rev, 2);
    assert_eq!(envelope.data.status, "applied");
}

#[test]
fn test_parse_ack_fixture() {
    let path = fixture_path("ack.json");
    let content = fs::read_to_string(&path).expect("Failed to read ack.json fixture");
    let envelope: Envelope<AckData> =
        serde_json::from_str(&content).expect("Failed to parse ack.json");

    assert_eq!(envelope.v, 1);
    assert_eq!(envelope.msg_type, "ack");
    assert_eq!(envelope.data.config_rev, 1);
    assert_eq!(envelope.data.accepted_seq, Some(1900));
    assert_eq!(envelope.data.persisted_seq, Some(1900));
}

#[test]
fn test_parse_error_fixture() {
    let path = fixture_path("error.json");
    let content = fs::read_to_string(&path).expect("Failed to read error.json fixture");
    let envelope: Envelope<ErrorData> =
        serde_json::from_str(&content).expect("Failed to parse error.json");

    assert_eq!(envelope.v, 1);
    assert_eq!(envelope.msg_type, "error");
    assert_eq!(envelope.data.code, "INSTANCE_MISMATCH");
}

File: worker/tests/contract.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { resolve } from 'path';
import {
  validateHelloPayload,
  validateReportPayload,
  AgentEnvelope,
  ServerEnvelope,
  HelloPayload,
  ReportPayload,
  WelcomeData,
  ConfigData,
  ConfigAckData,
  AckData,
  ErrorData,
} from '../src/protocol/types';

function loadFixture<T>(filename: string): T {
  const filePath = resolve(__dirname, '../../protocol/fixtures', filename);
  const raw = readFileSync(filePath, 'utf-8');
  return JSON.parse(raw) as T;
}

describe('Worker Protocol Contract Tests (Fixtures Alignment)', () => {
  it('correctly deserializes and validates hello.json fixture', () => {
    const envelope = loadFixture<AgentEnvelope<HelloPayload>>('hello.json');
    expect(envelope.v).toBe(1);
    expect(envelope.type).toBe('hello');
    expect(envelope.instance_id).toBe('550e8400-e29b-41d4-a716-446655440000');
    expect(envelope.seq).toBe(1);

    const valid = validateHelloPayload(envelope.data);
    expect(valid).toBe(true);
    expect(envelope.data.environment.type).toBe('container');
    expect(envelope.data.resources.cpu_capacity_cores).toBe(0.5);
  });

  it('correctly deserializes and validates report.json fixture', () => {
    const envelope = loadFixture<AgentEnvelope<ReportPayload>>('report.json');
    expect(envelope.v).toBe(1);
    expect(envelope.type).toBe('report');
    expect(envelope.data.samples).toBeDefined();
    expect(envelope.data.samples!.length).toBeGreaterThan(0);

    const valid = validateReportPayload(envelope.data);
    expect(valid).toBe(true);
    expect(envelope.data.samples![0].metrics.cpu.usage_pct).toBe(14.5);
    expect(envelope.data.samples![0].metrics.network.interface).toBe('eth0');
  });

  it('correctly deserializes welcome.json fixture', () => {
    const envelope = loadFixture<ServerEnvelope<WelcomeData>>('welcome.json');
    expect(envelope.v).toBe(1);
    expect(envelope.type).toBe('welcome');
    expect(envelope.data.config_rev).toBe(1);
    expect(envelope.data.config).toBeDefined();
    expect(envelope.data.config.sample_interval_sec).toBe(2);
  });

  it('correctly deserializes config.json fixture', () => {
    const envelope = loadFixture<ServerEnvelope<ConfigData>>('config.json');
    expect(envelope.v).toBe(1);
    expect(envelope.type).toBe('config');
    expect(envelope.data.config_rev).toBeDefined();
  });

  it('correctly deserializes config_ack.json fixture', () => {
    const envelope = loadFixture<AgentEnvelope<ConfigAckData>>('config_ack.json');
    expect(envelope.v).toBe(1);
    expect(envelope.type).toBe('config_ack');
    expect(envelope.data.config_rev).toBe(2);
    expect(envelope.data.status).toBe('applied');
  });

  it('correctly deserializes ack.json fixture', () => {
    const envelope = loadFixture<ServerEnvelope<AckData>>('ack.json');
    expect(envelope.v).toBe(1);
    expect(envelope.type).toBe('ack');
    expect(envelope.data.persisted_sample_seq).toBe(1001);
    expect(envelope.data.config_rev).toBe(1);
  });

  it('correctly deserializes error.json fixture', () => {
    const envelope = loadFixture<ServerEnvelope<ErrorData>>('error.json');
    expect(envelope.v).toBe(1);
    expect(envelope.type).toBe('error');
    expect(envelope.data.code).toBeDefined();
    expect(envelope.data.message).toBeDefined();
  });
});
